### DISCUSSION TITLE
WWDC18 image downsampling 및 prefecthAPI

### DISCUSSION THREAD

**Discussion by jeongdung-eo** [GEON-PPANG/GEON-PPANG-iOS#262] [General]
Bakery 리스트를 CollectionView에 무한 스크롤하여 이미지를 로딩하는 과정에서 이미지 로딩 속도가 느리고, 메모리 사용량이 급격하게 늘어나는 문제를 발견했는데요.
이를 해결하는 과정에서 알게된 [WWDC 2018 Image and Graphics Best Practices] 내용을 공유하고자 정리했습니다~,~

**[WWDC 2018 Image and Graphics Best Practices]**

 앱은 메모리와 CPU 같은 자원이 한정되어 있기 때문에, 배터리 수명과 앱의 반응성을 위해 메모리와 CPU를 따로 고려하지 않고 최적화할 방법을 찾아야한다.

 메모리 사용량은 이미지 파일의 크기가 아니라 이미지의 해상도로 결정된다. 즉, 메모리 사용량은 파일 크기가 아닌 이미지 크기(해상도)에 따라 달라진다.

**`UIImage와 UIImageView`**

**UIImage** 는 UIKit의 콘텐츠를 나타내며, **UIImageView** 는 UIKit에서 UIImage를 표시하는 데 사용됩니다.

<img width="1000" alt="스크린샷 2024-05-22 13 20 41" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/8649799e-2aae-4607-8662-6a1d6f218b6a">

MVC 패턴으로 생각하면, UIImage는 이미지를 로드하는 **Model** 이고, UIImageView는 렌더링된 이미지를 표시하는 **View** 라고 할 수 있습니다.

<img width="1000" alt="스크린샷 2024-05-22 13 21 28" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/8187b908-c289-43de-b65f-24d5e69b0dcb">

**`Image Rendering 프로세스`**

UIImage의 처리 방식은 크게 **Load** , **Decode** , **Render** 세 가지 프로세스로 진행됩니다.
<img width="935" alt="스크린샷 2024-05-22 13 14 52" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/844c1269-c036-4c57-891a-fce009d57b53">

각 단계에 대해서 살펴보면 

**`Load`**

- 압축된 이미지 파일(JPEG, PNG)을 메모리에 로드합니다.

**`Decode`**

- JPEG, PNG 등과 같은 이미지 파일을 GPU가 읽을 수 있도록 디코딩 작업을 실행합니다.
- 이 단계에서는 이미지 파일들을 이미지 버퍼로 디코딩합니다.

**`Render`**

- UIKit의 UIImageView에서 렌더링을 요청하면, 디코딩된 이미지 버퍼 데이터를 복사하여, 프레임 버퍼만큼 UIImageView의 크기에 맞춰 조정하는 작업을 실행합니다.
- 즉, 이미지 버퍼를 프레임 버퍼로 렌더링하여 UIImageView에 나타냅니다.

**`Buffer에 대한 이해`**

`Buffer`

- 버퍼는 메모리의 연속된 영역으로, 일반적으로 일련의 같은 크기의 요소로 간주됩니다.

<img width="998" alt="스크린샷 2024-05-22 13 23 27" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/f36fda01-9611-48b1-ae98-3308bacaac8c">

`image buffer`

- 이미지의 메모리 표현을 나타내는 버퍼입니다.
- 이미지 버퍼의 각 요소는 단일 픽셀의 색상 정보와 투명도를 나타내기 때문에 메모리의 버퍼 크기는 이미지 크기에 비례합니다.
- 따라서 더 큰 이미지를 표현하기 위해서는 더 많은 이미지 버퍼가 필요합니다.

<img width="995" alt="스크린샷 2024-05-22 13 26 23" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/6a5f255f-b80f-4eb0-b93a-d2870cae7d4e">

**`Frame Buffer`**

- 프레임 버퍼는 앱의 실제 렌더링된 결과를 보관하는 버퍼입니다.
- 앱이 뷰 계층 구조를 업데이트하면, UIKit은 앱의 윈도우와 모든 하위 뷰를 프레임 버퍼에 렌더링합니다.

<img width="995" alt="스크린샷 2024-05-22 13 25 45" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/bb923168-eeeb-49c9-999b-47641d073b7f">

- 프레임 버퍼는 디스플레이에 표시할 각 픽셀의 색상 정보를 제공하고, 디스플레이는 프레임 버퍼가 제공한 콘텐츠를 앱에 표시합니다.
- iPhone에서는 60Hz마다, iPad의 경우 120Hz마다 업데이트되어 디스플레이에 표시됩니다.

<img width="813" alt="스크린샷 2024-05-22 14 10 04" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/756e4293-3579-4d1e-8a38-c44caf5a6c79">

**`data buffer`**

- 데이터 버퍼는 이미지 버퍼의 다른 종류로, 일련의 데이터 바이트를 저장하는 버퍼입니다.
- 서버에서 다운로드되거나 디스크에 저장된 이미지 데이터를 저장합니다.
- 데이터 버퍼는 각 픽셀이 가진 색상과 투명도 정보가 없기 때문에 프레임 버퍼에 바로 적용할 수 없습니다. 인코딩된 이미지 데이터를 디코딩하여 픽셀의 색상과 투명도를 포함한 이미지 버퍼로 변환해야 합니다.

**`Data Buffer의 Rendering 과정`**

1. 로딩 준비
• 뷰에 로딩된 이미지 메타 데이터, UIImage, UIImageView를 준비합니다.

<img width="987" alt="스크린샷 2024-05-22 13 32 19" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/910747f0-b015-45a2-92db-2c3b250e92f4">

2. 픽셀 정보 얻기
    1. 이미지의 각 픽셀 정보를 프레임 버퍼에 채우려면 이미지 픽셀 정보를 가져와야 하는데, 이를 UIImage에서 처리합니다.
    2. UIImage는 디코딩 후 이미지 픽셀 정보를 저장하기 위해 이미지 크기만큼 이미지 버퍼를 할당합니다.

<img width="996" alt="스크린샷 2024-05-22 13 32 42" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/2e70ed28-2836-4239-9b09-0403ddb53152">

3. 화면에 표시
    1. UIKit이 UIImageView에 렌더링을 요청하면, 이미지 버퍼에 저장되어 있는 이미지 데이터(색상, 투명도)를 프레임 버퍼에 복사합니다.
    2. 표시할 디스플레이 크기에 맞게 조정하여 앱에 표시합니다.

<img width="995" alt="스크린샷 2024-05-22 13 30 46" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/7b4c1f2f-2b05-4d56-b579-c0ee5df3716e">

**`UIImageViewRendering 과정의 문제점`** 

<img width="815" alt="스크린샷 2024-05-22 15 03 49" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/673e4279-8ba6-407b-a2ef-3c3e1802f3ca">

**`UIImageView`** 는 이미지를 반복적으로 렌더링 요청하여, **`UIImage`** 가 이미지를 여러 번 디코딩하게 된. 이는 CPU 소모를 증가시킵니다.

이를 해결하기 위해, **`UIImage`** 는 이미지를 한 번만 디코딩하고 디코딩된 이미지 데이터를 이미지 버퍼에 보관하여 재사용합니다.

그러나, 디코딩된 이미지 버퍼를 메모리에 보관함으로써 메모리 사용량이 증가하고, 큰 이미지를 많이 사용할 경우 OS가 개입하여 메모리를 압축하게 됩니다.

이 과정에서 CPU 사용량이 증가하고, 최악의 경우 운영체제가 앱을 종료시킬 수 있습니다.

**💡 원본 이미지의 사이즈가 큰 경우에 UIImage가 디코딩된 이미지 버퍼를 가지고 있으므로 메모리 사용량이 증가합니다.**

### `Downsampling`

위에서 언급한 문제를 downsampling을 통해서 해결할 수 있습니다.

`기존의 프로세스`

<img width="997" alt="스크린샷 2024-05-22 13 41 28" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/cd1aa0a2-5810-4167-ae29-17e04590f394">

이미지가 로드되면 데이터 버퍼가 먼저 로드되고, 디코딩된 데이터인 이미지 버퍼가 복사되어 프레임 버퍼만큼 크기를 조절합니다.

**`Downsampling의 프로세스`**

<img width="987" alt="스크린샷 2024-05-22 13 43 39" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/0fbb1c7c-e22d-41d7-b66e-847487da8790">

1. 원본 이미지를 UIImageView의 사이즈를 고려하여 렌더링을 통해 축소하여 썸네일을 만듭니다.
2. 썸네일을 만든 뒤, 데이터 버퍼를 삭제하여 메모리 사용량을 줄입니다.
3. 썸네일을 디코딩하면, 할당되는 이미지 버퍼의 크기도 줄어듭니다.

   💡 디코딩하기 전에 데이터 버퍼 자체의 사이즈를 줄이면, 이미지 버퍼의 크기가 줄어들기 때문에 이미지 로딩에 사용되는 전체 메모리 크기도 줄어듭니다. 이를 통해 최소한의 메모리 사용량을 유지할 수 있습니다.

이를 통해 각 셀의 이미지 표시를 최적화하여 전체 메모리 사용량을 크게 줄일 수 있어 유용합니다. 그러나 이로 인해 또 다른 문제가 발생할 수 있습니다.

뷰가 스크롤되면 CPU가 프레임 버퍼에 표시할 콘텐츠를 빠르게 렌더링할 수 있지만, 스크롤이 너무 빠르면 CPU는 여전히 Core Graphics에 참여해야 합니다. 프레임을 해석하는 데 시간이 많이 걸리므로, 하드웨어가 프레임 버퍼 데이터를 읽어 화면에 표시할 때 프레임 버퍼 데이터가 준비되지 않아 사용자 관점에서 지연이 발생할 수 있습니다. 또한, CPU가 불안정할 경우 배터리 수명에 영향을 줄 수 있습니다. 

이 문제를 **Prefetching** 과 **Background Decoding/Downsampling** 을 통해 해결할 수 있습니다.

### **`Prefetching과 Background Decoding/Downsampling`**

<img width="995" alt="스크린샷 2024-05-22 13 49 30" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/88e1bacf-74cf-4f87-94f5-63ba6bd08fc5">

UICollectionView에서 무한 스크롤을 구현하는 경우, 다음과 같이 구현할 수 있습니다.

1. **Prefetching** 된 셀에 대해 **`DispatchQueue.global`** 에서 이미지 디코딩을 수행합니다.
<img width="1001" alt="스크린샷 2024-05-22 13 49 46" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/036b650f-3d1e-4444-8453-59f249a22f71">

그러나 이러한 경우 **thread explosion** 이 발생할 수 있습니다. 한 번에 여러 이미지를 처리하지만, 디바이스에 CPU가 두 개만 있는 경우, GCD는 디코딩 작업을 처리하기 위해 새로운 스레드를 생성하게 됩니다. 새로운 스레드를 생성하고 다른 스레드 간에 **context switching** 을 하는 것은 시간과 리소스를 많이 소모합니다.

1. 이를 해결하기 위해 디코딩 작업을 **serialQueue** 에 **async** 하게 전달할 수 있습니다.

<img width="999" alt="스크린샷 2024-05-22 13 51 06" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/dcf3cd00-3cf5-407e-a3b5-1107270d416d">

이로 인해 일부 이미지의 디코딩 프로세스가 지연될 수 있지만, 더 중요한 것은 **thread context switching** 에서 낭비되는 시간과 리소스를 줄일 수 있다는 점입니다.

**[Downsampling 코드 분석]**
<img width="750" alt="스크린샷 2024-05-23 13 56 34" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/b631b417-3cf5-4bfb-a1a3-8c4b67d590ae">

1. CGImageSource 객체 생성
<img width="650" alt="스크린샷 2024-05-23 13 56 50" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/220d3ae7-17d8-4967-8e51-8af1163ef182">

  
    - `kCGImageSourceShouldCache` 는 이미지를 디코딩된 형식으로 캐싱할지 설정 하는 옵션입니다.
    - 데이터 버퍼는 Downsampling 이후 제거할 것이기에 KCGImageSourceShouldCache를 false로 설정합니다.
2. 화면에 렌더링해야하는 실제 크기 계산
    
<img width="556" alt="스크린샷 2024-05-23 13 57 06" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/c103f310-b885-4baf-a625-8a740a44713a">

    - scale과 렌더링할 UIImageView 사이즈에 맞춰 썸네일의 최대 크기를 계산합니다.
3. 썸네일 얻기
    
<img width="1018" alt="스크린샷 2024-05-23 14 23 00" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/3ac3acb6-42d6-412b-a621-12034cdfa61b">

    - `kCGImageSourceCreateThumbnailFromImageAlways` 는 이미지 원본 파일에 썸네일이 있어도 전체 이미지를 이용해 썸네일을 만들지 결정정합니다.  `mageSourceThumbnailMaxPixelSize`를 지정하지 않으면 썸네일의 크기는 전체 이미지 크기가 됩니다. 위 코드에서는 항상 썸네일을 만들도록 설정했습니다.
    - `kCGImageSourceShouldCacheImmediately` 는 썸네일을 생성할 때 이미지 버퍼를 생성하라고 알려줍니다. 이 옵션이 가장 중요한데 Core Graphics에게 지금 썸네일이 생성되었으니, 디코딩된 이미지 버퍼를 생성하라고 알려주기 때문입니다. 이 옵션을 통해 디코딩에 쓰이는 CPU Hit 순간을 정확히 제어할 수 있습니다.
    - `kCGImageSourceCreateThumbnailWithTransform` : 원본 이미지의 방향 및 비율에 맞게 썸네일을 회전하고 scaling 할지 결정하는 옵션입니다.
    - `kCGImageSourceThumbnailMaxPixelSize` : 썸네일 이미지의 최대 가로, 세로입니다. point가 아니라 픽셀 단위로 지정해야 합니다. 이 옵션을 지정하지 않으면 썸네일의 크기가 원본 이미지 크기와 동일하게 설정됩니다.
<img width="1031" alt="스크린샷 2024-05-23 14 23 41" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/107853954/890af40e-e235-4ead-8d1c-f17617554988">

    - 마지막으로 `CGImage` 타입으로 downsampling한 썸네일 이미지를 생성합니다.
    - 최종적으로는 UIImage로 쓰기 위해 변환을 해주고 리턴을 해줍니다.

[참고]

https://jeong9216.tistory.com/560#uiimage-and-uiimageview

https://velog.io/@dohyun8032/WWDC-2018-Image-and-Graphics-Best-Practices

https://docs.huihoo.com/apple/wwdc/2018/219_image_and_graphics_best_practices.pdf

**Comment by seongmin221** [comment]
굉장히 어렵군요 .... 두고두고 읽어봐야겠습니닷
짱좋은 자료 만들어줘서 감사티비 !!

kingfisher에서는 이걸 어떻게 처리하는지 확인하고 싶어서 함 안에 들여다보니까
먼저 `.setImage(...)`는 내부에 이런 내용들이 있었습니닷

<img width="867" alt="image" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/72431640/183f7942-674c-4ca8-8544-0ee6cbb83f6d">

<br>

그래서 `KingfisherOptionsInfoItem` 으로 들어가보니까 `enum`으로 선언되어 있고, 내부에 저희가 사용할 `processor`가 선언되어 있더라구요 !

<img width="822" alt="image" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/72431640/28404d58-e31c-496c-ba16-24e08ac75f48">

<br>

`processor` 의 입력값으로 들어가야 할 `ImageProcessor` 는 아래와 같은 프로토콜로 선언되어 있습니다 !

<img width="970" alt="image" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/72431640/7835ec61-74a7-4aea-9326-696889e505cd">

여기에는 신기한 프로퍼티가 있는데, 바로 `identifier` 입니닷

읽어보니까 caching 하고 caching 된 이미지를 가져올 때 이 identifier로 processor를 판단 한다네요 ??
사실 무슨 소리인지는 잘 모르겠다만 ... caching 할 때에는 동일한 identifier 쓰라는듯....?
저희 다운샘플링에서는 아직은 필요가 없지 않을까 싶습니다 ~

<br>

무튼 실제 다운샘플링이 실행되는 코드를 보니까 아래와 같습니닷

<img width="1056" alt="image" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/72431640/b23d9f14-83f1-4ea1-82a3-beb0e2c8d7a7">

사실 Kingfisher 쪽 코드 자체가 막 이해가 잘 되지는 않지만 ... 다운샘플링 개념이 아직 부족해서 ... ^___^
대강 이해되는 바로는 지정한 사이즈에 맞게 Thumbnail로 만드는 것 같긴 합니다 !!

신기방기 ....

킹피셔 코드 분석도 해주세요 ^________^

**Reply by jeongdung-eo** [reply]
kingfisher 내부에서 downsampling하는 로직이 WWDC18에서 보여준 코드와 동일하여, 위의 코드 분석 내용에 추가하겠습니다 :)
+ kingfisher cache 과정을 이해하면서 참고했던 블로그가 있는데 공유하겠슴당~,~
https://dev200ok.blogspot.com/2020/09/ios-kingfisher.html
